3.0/m_roleplay.cpp

```diff
@@ -0,0 +1,704 @@
+		// of these are restricted at the protocol level we only need to check for SOH
+		// and SPACE.
+		return (text.length() >= 2) && (text[0] == '\x1') &&  (text[1] != '\x1') && (text[1] != ' ');
+	}
+};
+
+// This is here to make the channel mode optional and configurable.
+class RoleplayMode : public SimpleChannelModeHandler
+{
+public:
+	RoleplayMode(Module* Creator)
+		: SimpleChannelModeHandler(Creator, "roleplay", '\0')
+	{
+		ConfigTag* tag = ServerInstance->Config->ConfValue("roleplay");
+		mode = tag->getString("mode", "U", 1, 1)[0];
+		if(!tag->getBool("needchanmode", true))
+			DisableAutoRegister();
+	}
+};
+
+/* This class does the heavy lifting of handling all the sending machinery. It
+ * helps cut back heavily on code duplication.
+ */
+class CommandBaseRoleplay : public Command
+{
+	SimpleChannelModeHandler& roleplaymode;
+
+	bool CheckMessage(User* user, MessageTarget& msgtarget, MessageDetailsImpl& msgdetails)
+	{
+		// Bypass restrictions/checks if this permission is active
+		if(!user->HasPrivPermission("channels/roleplay-override"))
+		{
+			// Don't allow CTCP spoofing.
+			std::string ctcpname;
+			if(msgdetails.IsCTCP(ctcpname) && !irc::equals(ctcpname, "ACTION"))
+				return false;
+
+			/* Inform modules that a message wants to be sent and check the
+			 * results. This stops ban/censor evasion, etc. To do this we
+			 * simply present this as a message from the user (effectively,
+			 * that's what this is).
+			 *
+			 * This works just like the privmsg logic (I stole it from
+			 * there. :p)
+			 *
+			 * --Elizafox
+			 */
+			ModResult modres;
+
+			FIRST_MOD_RESULT(OnUserPreMessage, modres, (user, msgtarget, msgdetails));
+			if (modres == MOD_RES_DENY)
+			{
+				// Inform modules that a module blocked the mssage.
+				FOREACH_MOD(OnUserMessageBlocked, (user, msgtarget, msgdetails));
+				return false;
+			}
+
+			// Check whether a module zapped the message body.
+			if (msgdetails.text.empty())
+			{
+				user->WriteNumeric(ERR_NOTEXTTOSEND, "No text to send");
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	bool CheckChannelPermissions(User* user, Channel* c)
+	{
+		if(!c->HasUser(user))
+		{
+			user->WriteNumeric(ERR_NOTONCHANNEL, c->name, "You're not on that channel");
+			return false;
+		}
+
+		// Ignore restrictions if the operator has channels/roleplay-override
+		if(!user->HasPrivPermission("channels/roleplay-override"))
+		{
+			if(need_op && c->GetPrefixValue(user) < OP_VALUE)
+			{
+				user->WriteNumeric(ERR_CHANOPRIVSNEEDED, c->name, "You're not a channel operator");
+				return false;
+			}
+
+			if(need_mode && !c->IsModeSet(roleplaymode))
+			{
+				user->WriteNumeric(ERR_ROLEPLAY, c->name, InspIRCd::Format("Channel mode +%c must be set", roleplaymode.GetModeChar()));
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	inline void SendMessage(User* user, Channel* c, const std::string& source, MessageTarget& msgtarget, MessageDetails& msgdetails)
```

Comment:
`inline` in the way it is used in this module is largely unnecessary because the functions are not exposed as part of the module's interface so the compiler can do whatever it wants with it.